velox/type/Filter.cpp

```diff
@@ -458,6 +472,55 @@ bool MultiRange::testBytesRange(
   return false;
 }
 
+std::unique_ptr<Filter> MultiRange::mergeWith(const Filter* other) const {
+  switch (other->kind()) {
+    case FilterKind::kMultiRange: {
+      const MultiRange* multiRangeOther = static_cast<const MultiRange*>(other);
+      bool bothNullAllowed = nullAllowed_ && other->testNull();
+      bool bothNanAllowed = nanAllowed_ && multiRangeOther->nanAllowed_;
+      std::vector<std::unique_ptr<Filter>> merged;
+      for (auto const& filter : this->filters()) {
+        for (auto const& filterOther : multiRangeOther->filters()) {
+          auto innerMerged = filter->mergeWith(filterOther.get());
+          switch (innerMerged->kind()) {
+            case FilterKind::kAlwaysFalse:
+            case FilterKind::kIsNull:
+              break;
+            default:
+              merged.push_back(std::move(innerMerged));
+          }
+        }
+      }
+
+      if (merged.empty()) {
+        return nullOrFalse(bothNullAllowed);
+      } else if (merged.size() == 1) {
+        return merged[0]->clone(bothNullAllowed);
+      } else {
+        return std::unique_ptr<Filter>(
+            new MultiRange(std::move(merged), bothNullAllowed, bothNanAllowed));
+      }
+    }
+
+    // Rules of MultiRange with IsNull/IsNotNull
+    // 1. MultiRange(nullAllowed=true) AND IS NULL => IS NULL
+    // 2. MultiRange(nullAllowed=true) AND IS NOT NULL =>
+    // MultiRange(nullAllowed=false)
+    // 3. MultiRange(nullAllowed=false) AND IS NULL
+    // => ALWAYS FALSE
+    // 4. MultiRange(nullAllowed=false) AND IS NOT NULL
+    // =>MultiRange(nullAllowed=false)
+    case FilterKind::kAlwaysTrue:
+    case FilterKind::kAlwaysFalse:
+    case FilterKind::kIsNull:
+      return other->mergeWith(this);
+    case FilterKind::kIsNotNull:
+      return this->clone(/*nullAllowed=*/false);
+    default:
+      VELOX_UNREACHABLE();
```

Comment:
Looks like the remaining things to add here would be:

kDoubleRange
kFloatRange
kBytesRange
kBytesValues

These can come in a follow-up PR though.

One ask here is to change the order of the 'case' statements to match all other implementations to allow for easier side-by-side comparison, e.g.

    case FilterKind::kAlwaysTrue:
    case FilterKind::kAlwaysFalse:
    case FilterKind::kIsNull:
      return other->mergeWith(this);
    case FilterKind::kIsNotNull:
      return this->clone(/*nullAllowed=*/false);
    case FilterKind::kDoubleRange:
    case FilterKind::kFloatRange:
    case FilterKind::kBytesRange:
    case FilterKind::kBytesValues:
      // TODO Implement
      VELOX_UNREACHABLE();
    case FilterKind::kMultiRange:
      ...
    default:
      VELOX_UNREACHABLE();

Reply:
It makes sense. Done.